[ci] Add riscv opt-int build

zhangfeiv0 opened the pull request:
Hi, @malfet 
Based on the previous discussion:

[RISCV CI support · Issue #141550 · pytorch/pytorch](https://github.com/pytorch/pytorch/issues/141550)

I have cross-compiled PyTorch for the RISC-V architecture on x86_64 Ubuntu 24.04 and created a new PR for it. Could you please help review it?


cc @malfet @seemethere

zhangfeiv0 committed:
ci: add riscv support

Signed-off-by: Zhang fei <zhangfei@iscas.ac.cn>

zhangfeiv0 commented:
Hi, @malfet 

Here’s a gentle ping: we have fixed some code errors detected by the CI. When you have time, could you please take a look and provide any suggestions?

zhangfeiv0 commented:
@Skylion007 Could you help review it if you have time?

wangfanv0 commented:
@malfet @Skylion007

malfet commented:
@zhangfeiv0 Do you mind changing it to periodic and than it's probably ok to be landed
Also you'll have to add ciflow/riscv label name to list of recognized labels (actually let me do it now, to avoid chicken and egg problem with our CI setup)

malfet commented:
@pytorchbot rebase

malfet commented on .github/workflows/riscv64.yml:
```diff
@@ -0,0 +1,26 @@
+name: riscv64
+
+on:
+  push:
+    branches:
+      - main
```
```suggestion
  pull_request:
```

malfet commented on .github/workflows/riscv64.yml:
```diff
@@ -0,0 +1,24 @@
+name: riscv64
+
+on:
+  pull_request:
+    tags:
```
```suggestion
    pull_request:
    tags:
```

malfet commented on .github/workflows/riscv64.yml:
```diff
@@ -0,0 +1,24 @@
+name: riscv64
+
+on:
+    pull_request:
+    tags:
```
```suggestion
  pull_request:
  push:
    tags:
```

malfet commented:
Making some changes to the workflow now. If it finished build in reasonable amount of time, than yes, it could be landed as an opt-in for now

malfet commented:
@zhangfeiv0, you'll need to adapt your PR to make it work with CI build, i.e. docker for CI is build using files in `.ci/docker` folder, rather than in `.ci/docker/manywheel`, and are stored on ECR/ghcr.io rather than on docker hub

zhangfeiv0 commented:
@malfet  I’m sorry, but I’m not very familiar with the PyTorch CI process. Could you please clarify if you mean using .ci/docker/build.sh to build the Docker image? And which specific workflow is used to trigger .ci/docker/build.sh?

zhangfeiv0 commented:
@malfet  Do you have time to continue reviewing?

lamaalrajih commented:
Hi @malfet. Could you review this PR? Thanks!

zhangfeiv0 commented:
@malfet  Could you please reopen the workflow for me? Thank you so much!

The Docker image build failed on linux.12xlarge, but it worked fine on [ubuntu-latest previously](https://github.com/pytorch/pytorch/actions/runs/13433721332/job/37531169923). I'm not sure if the issue is due to differences in the environment, so I've switched the runner back to ubuntu-latest. Could you share the OS environment details of linux.12xlarge? This way, if errors occur in the future, I can test them locally.

malfet commented:
@pytorchbot rebase

zhangfeiv0 commented:
@malfet  First of all, thank you for enabling the workflow for me.  

An error occurred during cross-compilation:  
```
/bin/sh: 1: /var/lib/jenkins/workspace/build/bin/protoc-3.13.0.0: Exec format error
```

To address this, I have already used `docker/setup-qemu-action@v3` to allow the host machine to run RISC-V programs during the compilation phase. However, this doesn’t seem to have worked. I’m not entirely sure if this is related to the runner environment. Could you provide some suggestions?

malfet commented:
@pytorchbot rebase

malfet commented:
> An error occurred during cross-compilation:
> 
> ```
> /bin/sh: 1: /var/lib/jenkins/workspace/build/bin/protoc-3.13.0.0: Exec format error
> ```
> 

I suggest disabling ONNX for now, which is the only component that requires protobuf. Let me try that, after another round of rebase, which should render sleef updates unnecessary.

malfet commented on .ci/pytorch/build.sh:
```diff
@@ -99,6 +99,38 @@ if [[ "$BUILD_ENVIRONMENT" == *aarch64* ]]; then
+  export USE_DISTRIBUTED=0
+  export USE_MKLDNN=0
+
+  # Reference: https://github.com/shibatch/sleef/tree/master/docs/1-user-guide#method-2
+  export SLEEF_TARGET_EXEC_USE_QEMU=ON
+  # TODO: Remove the following sleef update code once the submodule is updated to the b56eab1
+  # Update sleef submodule for RISC-V
+  sudo chown -R jenkins /var/lib/jenkins/workspace /opt
+  git config --global --add safe.directory /var/lib/jenkins/workspace
+  git submodule update --init third_party/sleef
+  cd third_party/sleef
+  git remote add upstream https://github.com/shibatch/sleef || true
+  git fetch upstream
+  git checkout master
+  git pull upstream "$(git rev-parse --abbrev-ref HEAD)"
+  cd ../..
```
```suggestion
```

malfet commented on .ci/pytorch/build.sh:
```diff
@@ -99,6 +99,38 @@ if [[ "$BUILD_ENVIRONMENT" == *aarch64* ]]; then
+
+  export CMAKE_CROSSCOMPILING=TRUE
+  export CMAKE_SYSTEM_NAME=Linux
+  export CMAKE_SYSTEM_PROCESSOR=riscv64
```
IMO if one really cross-compiles, it should be `CMAKE_TARGET_PROCESSOR` rather than a system one

malfet commented on .ci/pytorch/common_utils.sh:
```diff
@@ -42,7 +42,8 @@ declare -f -t trap_add
     # TODO: we should add an option to `build_amd.py` that reverts the repo to
     #       an unmodified state.
-    if [[ "$BUILD_ENVIRONMENT" != *rocm* ]] && [[ "$BUILD_ENVIRONMENT" != *xla* ]] ; then
+    # TODO: Remove riscv once the sleef submodule is updated to the b56eab1
+    if [[ "$BUILD_ENVIRONMENT" != *rocm* ]] && [[ "$BUILD_ENVIRONMENT" != *xla* ]] && [[ "$BUILD_ENVIRONMENT" != *riscv* ]]; then
```
```suggestion
    if [[ "$BUILD_ENVIRONMENT" != *rocm* ]] && [[ "$BUILD_ENVIRONMENT" != *xla* ]] ; then
```

malfet committed:
Apply suggestions from code review

malfet commented on .ci/pytorch/build.sh:
```diff
@@ -99,6 +99,25 @@ if [[ "$BUILD_ENVIRONMENT" == *aarch64* ]]; then
+
+  export USE_CUDA=0
+  export USE_DISTRIBUTED=0
+  export USE_MKLDNN=0
```
```suggestion
  export USE_MKLDNN=0
  export BUILD_CUSTOM_PROTOBUF=0
```

malfet committed:
Update .ci/pytorch/build.sh

zhangfeiv0 committed:
ci: Update

Signed-off-by: Zhang fei <zhangfei@iscas.ac.cn>

zhangfeiv0 commented:
During cross-compilation, some third-party libraries need to run on the host machine, including `protocbuf` and `sleef`.  

- **Protobuf**: As per your suggestion, it was disabled.  
- **Sleef**: Still requires execution on the host machine during compilation.  

To address this, I introduced `docker/setup-qemu-action@v3` to emulate the execution of the necessary binaries in these third-party libraries. This approach was also tested in my local environment. However, in the last workflow run, the `protobuf` binary failed to execute, suggesting this solution did not work as intended.  
I moved this step into the **Build step** to see if it resolves the issue. If successful, this would eliminate the need to disable `protobuf`.  

Additionally, some steps you previously deleted need to be retained; otherwise, compilation errors will still occur during the build phase.  

@malfet Could you restart the workflow and check again? Thanks!

zhangfeiv0 committed:
ci: Update

Signed-off-by: Zhang fei <zhangfei@iscas.ac.cn>

lamaalrajih commented:
@malfet bump!

zhangfeiv0 committed:
ci: Update

Signed-off-by: Zhang fei <zhangfei@iscas.ac.cn>

zhangfeiv0 commented:
@houseroad Friendly ping

zhangfeiv0 committed:
ci: Update

Signed-off-by: Zhang fei <zhangfei@iscas.ac.cn>

zhangfeiv0 committed:
ci: Update

Signed-off-by: Zhang fei <zhangfei@iscas.ac.cn>

zhangfeiv0 committed:
ci: Update

Signed-off-by: Zhang fei <zhangfei@iscas.ac.cn>

zhangfeiv0 committed:
ci: Update

Signed-off-by: Zhang fei <zhangfei@iscas.ac.cn>

zhangfeiv0 committed:
ci: Update

Signed-off-by: Zhang fei <zhangfei@iscas.ac.cn>

zhangfeiv0 committed:
ci: Update

Signed-off-by: Zhang fei <zhangfei@iscas.ac.cn>

malfet commented on .github/workflows/riscv64.yml:
```diff
@@ -0,0 +1,25 @@
+name: riscv64
+
+on:
+  pull_request:
```
This is just for testing, right? I.e. at the beginning it should be an opt-in workflow, that later can be changed to periodic?

zhangfeiv0 replied:
Yes, I removed the pull_request trigger.

malfet commented on .ci/docker/ubuntu-riscv/Dockerfile:
```diff
@@ -0,0 +1,154 @@
+ARG UBUNTU_VERSION
```
Please add a comment that this is a cross-compiled docker container and may be re-name folder to `ubuntu-cross-riscv` or something like that

malfet commented on .github/workflows/_linux-build.yml:
```diff
@@ -287,10 +287,36 @@ jobs:
+            # EC2 specific setup for RISC-V emulation
+            # Ensure binfmt_misc is available
+            echo "Mounting binfmt_misc filesystem"
+            sudo mount binfmt_misc -t binfmt_misc /proc/sys/fs/binfmt_misc 2>/dev/null || true
+
+            echo "QEMU registration: multiarch/qemu-user-static"
+            docker run --rm --privileged multiarch/qemu-user-static --reset -p yes || true
```
Hmm, why all of this is needed when one cross-compiles? Wouldn't defining CMAKE_TARGET_SYSTEM=riscv-XYZ be suficient?

zhangfeiv0 replied:
Because some dependency packages still need to run RISC-V target code on the host machine during cross-compilation—for example, when cross-compiling SLEEF, it requires running binary script files like `mkrename`—and to compile for the target architecture, the package for the host architecture must first be built, which is quite cumbersome. Therefore, `qemu-user-static` is installed to avoid these issues.

zhangfeiv0 committed:
ci: Update

Signed-off-by: Zhang fei <zhangfei@iscas.ac.cn>

malfet reviewed (approved):
LGTM, thank you for all the updates

malfet commented:
@pytorchbot merge -f "This is opt-in CI PR, and lint is green"
